data_adapter_oemof/build_datapackage.py

```diff
@@ -121,38 +151,50 @@ def build_datapackage(cls, adapter: Adapter):
             facade_adapter_name: str = PROCESS_TYPE_MAP[process_name]
             facade_adapter = FACADE_ADAPTERS[facade_adapter_name]
             components = []
+            process_busses = []
             # Build class from adapter with Mapper and add up for each component within the Element
             for component_data in process_data.scalars.to_dict(orient="records"):
+                component_mapper = Mapper(
+                    adapter=facade_adapter,
+                    process_name=process_name,
+                    data=component_data,
+                    timeseries=timeseries,
+                )
                 component = facade_adapter.parametrize_dataclass(
-                    process_name, component_data, timeseries, struct
+                    struct, component_mapper
                 )
                 components.append(component.as_dict())
-                component_mapper = Mapper(
-                        adapter=facade_adapter,
-                        process_name=process_name,
-                        data=component_data,
-                        timeseries=timeseries,
-                    )
-                # Fill bus.csv with all busses occurring
-                parametrized_elements["bus"] += list(
-                    component_mapper
-                    .get_busses(struct)
-                    .values()
+                # Fill with all busses occurring, needed for foreign keys as well!
+                process_busses += list(component_mapper.get_busses(struct).values())
+
+            process_busses = pd.unique(process_busses)
+            parametrized_elements["bus"] += process_busses
+
+            # check if facade_adapter already exists in foreign keys -> add or update accordingly
+            if facade_adapter_name in foreign_keys.keys():
+                # The same column cannot be pointing to different locations nor be None for some entries.
+                assert foreign_keys[facade_adapter_name] == cls.get_foreign_keys(
+                    process_busses, component_mapper, components
+                ), "Foreign keys have to be the equal for every instance of the FacadeAdapter. "
+            else:
+                foreign_keys[facade_adapter_name] = cls.get_foreign_keys(
+                    process_busses, component_mapper, components
```

Comment:
`get_foreign_keys` must be placed in for loop over components, as `component_mapper` can differ within a process!

Reply:
But we can only have one foreign key set for one Process.
We could use this as a check and then seperate the two Processes into two processes once they differ?

Reply:
We have to distinguish here:
we can have multiple foreign keys per process. What we cannot have is mixed columns, containing scalars and FKs to sequences (aka for one process we use a scalar efficiency, for the other one we use a timeseries as efficiency).
But the way, the mapper currently works - this isn't even possible.
So I think we are good here.
Having this in mind - you were right - foreign keys can be set once per process, as different FKs per row woudl result in corrupt datapacakge anyway!